pkg/services/live/pipeline/converter.go

```diff
@@ -0,0 +1,24 @@
+package pipeline
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+// ChannelFrame is a wrapper over data.Frame with additional channel information.
+// Channel is used for rule routing, if the channel is empty then frame processing
+// will try to take current rule Processor and Outputter. If channel is not empty
+// then frame processing will be redirected to a corresponding channel rule.
+// TODO: avoid recursion, increment a counter while frame travels over pipeline steps, make it configurable.
+type ChannelFrame struct {
+	Channel string
+	Frame   *data.Frame
+}
```

Comment:
Maybe:
```suggestion
type ChannelMessage struct {
	Channel string
	Frame   *data.Frame
	Body     interface{}
}
```

We still need some way to keep the original JSON messages and pass them though to the front end, while also converting to frames and processing.

Does that break your model?

Reply:
Currently does not fit well interfaces. This is an extra field to manually maintain every time inside every entity, also when returning  ChannelFrame from outputs. And we need to modify Processor interface.

I think a better place for global pipeline things is `Vars` currently.